src/delegation/payload.rs

```diff
@@ -0,0 +1,488 @@
+use super::policy::{predicate, Predicate};
+use crate::ability::arguments::Named;
+use crate::time;
+use crate::{
+    capsule::Capsule,
+    crypto::{varsig, Nonce},
+    did::{Did, Verifiable},
+    time::{TimeBoundError, Timestamp},
+};
+use core::str::FromStr;
+use derive_builder::Builder;
+use did_url::DID;
+use libipld_core::{codec::Codec, error::SerdeError, ipld::Ipld, serde as ipld_serde};
+use serde::{Deserialize, Serialize};
+use std::{collections::BTreeMap, fmt::Debug};
+use thiserror::Error;
+use web_time::SystemTime;
+
+#[cfg(feature = "test_utils")]
+use proptest::prelude::*;
+
+#[cfg(feature = "test_utils")]
+use crate::ipld;
+
+/// The payload portion of a [`Delegation`][super::Delegation].
+///
+/// This contains the semantic information about the delegation, including the
+/// issuer, subject, audience, the delegated ability, time bounds, and so on.
+#[derive(Debug, Clone, PartialEq, Builder)] // FIXME Serialize, Deserialize, Builder)]
+pub struct Payload<DID: Did> {
+    /// The subject of the [`Delegation`].
+    ///
+    /// This role *must* have issued the earlier (root)
+    /// delegation in the chain. This makes the chains
+    /// self-certifying.
+    ///
+    /// The semantics of the delegation are established
+    /// by the subject.
+    ///
+    /// [`Delegation`]: super::Delegation
+    pub subject: Option<DID>,
+
+    /// The issuer of the [`Delegation`].
+    ///
+    /// This [`Did`] *must* match the signature on
+    /// the outer layer of [`Delegation`].
+    ///
+    /// [`Delegation`]: super::Delegation
+    pub issuer: DID,
+
+    /// The agent being delegated to.
+    pub audience: DID,
+
+    /// A [`Did`] that must be in the delegation chain at invocation time.
+    #[builder(default)]
+    pub via: Option<DID>,
+
+    /// The command being delegated.
+    pub command: String,
+
+    /// Any [`Predicate`] policies that constrain the `args` on an [`Invocation`][crate::invocation::Invocation].
+    #[builder(default)]
+    pub policy: Vec<Predicate>,
+
+    /// Extensible, free-form fields.
+    #[builder(default)]
+    pub metadata: BTreeMap<String, Ipld>,
+
+    /// A [cryptographic nonce] to ensure that the UCAN's [`Cid`] is unique.
+    ///
+    /// [cryptograpgic nonce]: https://en.wikipedia.org/wiki/Cryptographic_nonce
+    /// [`Cid`]: libipld_core::cid::Cid ;
+    #[builder(default = "Nonce::generate_16(&mut vec![])")]
+    pub nonce: Nonce,
+
+    /// The latest wall-clock time that the UCAN is valid until,
+    /// given as a [Unix timestamp].
+    ///
+    /// [Unix timestamp]: https://en.wikipedia.org/wiki/Unix_time
+    pub expiration: Timestamp,
```

Comment:
```suggestion
    pub expiration: Option<Timestamp>,
```

Shouldn't this be optional? It *must* be set to `null`, if not present, but `Timestamp` doesn't have a representation for `null` I think.

> The exp field MUST be set. Following the [principle of least authority](https://en.wikipedia.org/wiki/Principle_of_least_privilege), it is RECOMMENDED to give a timestamp expiry for UCANs. If the token explicitly never expires, the exp field MUST be set to null. If the time is in the past at validation time, the token MUST be treated as expired and invalid.